### PR TITLE
Adds test for phpcs using Matomo coding standards, #PG-3751

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

Adds test for phpcs using Matomo coding standards
Fixes: #PG-3751

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Review by snake14 — COMMENTED** [review]
Looking really good :+1:  The only concern I have is the code sniffer dependency (vendor/) being committed. That doesn't seem like something that should be included in production.

**Comment by AltamashShaikh** [comment]
> Looking really good 👍 The only concern I have is the code sniffer dependency (vendor/) being committed. That doesn't seem like something that should be included in production.

@snake14 Slevomat requires [phpcodensiffer](https://github.com/slevomat/coding-standard/blob/master/composer.json#L15)

**Comment by snake14** [comment]
> @snake14 Slevomat requires [phpcodensiffer](https://github.com/slevomat/coding-standard/blob/master/composer.json#L15)

@AltamashShaikh Shouldn't Slevomat be `require-dev` too?

**Comment by AltamashShaikh** [comment]
@snake14 As discussed I `gitignored` all those libraries

**Comment by snake14** [comment]
> @snake14 As discussed I `gitignored` all those libraries

Thank you. That looks a lot better. You could probably ignore the two new files in `vendor/bin/` as well, but it's not a big deal.

**Review by snake14 — APPROVED** [review]

**Comment by AltamashShaikh** [comment]
> > @snake14 As discussed I `gitignored` all those libraries
> 
> Thank you. That looks a lot better. You could probably ignore the two new files in `vendor/bin/` as well, but it's not a big deal.

@snake14 Done :)

File: phpcs.xml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+<?xml version="1.0"?>
```
Comment (sgiehl) [inline phpcs.xml:1]: This files needs to be reworked, as it should now use our new coding standard. Similar to https://github.com/matomo-org/matomo/blob/5.x-dev/phpcs.xml, but you need to adjust it for the plugin accordingly.

File: composer.json [hunk 2]
```diff
@@ -27,5 +36,8 @@
             "AnalyticsReporting",
             "Oauth2"
         ]
+    },
+    "require-dev": {
+        "matomo-org/matomo-coding-standards": "dev-master"
```
Comment (sgiehl) [inline composer.json]: I'm not sure if requiring the coding standard directly in the repo is useful. We have a lot of plugins that don't even use composer, so for them that would be a total overhead.
We could also simply run a `composer require` for the package within the github action to ensure its available there. (Note: We might need to publish the coding style on packagist to be able to require it on command line)

**Review by sgiehl — CHANGES_REQUESTED** [review]

File: composer.json (hunk 2 above)
Comment (AltamashShaikh) [reply composer.json]: I will try without adding in composer and see how does it look

**Comment by michalkleiner** [comment]
I think it would be great to split this PR into two - one where we add the config and any necessary dependencies, and leave the reformatting for a separate one. Together it makes it harder to review.

File: composer.json [hunk 3]
```diff
@@ -12,7 +12,10 @@
         "phpseclib/bcmath_compat": "^2.0"
     },
     "config": {
-        "platform-check": "false"
+        "platform-check": "false",
+        "allow-plugins": {
+            "dealerdirect/phpcodesniffer-composer-installer": false
```
Comment (sgiehl) [inline composer.json]: I guess the changes to composer files can be fully reverted.

Comment (AltamashShaikh) [reply composer.json]: Yes :+1:

File: phpcs.xml (hunk 1 above)
Comment (AltamashShaikh) [reply phpcs.xml:1]: Done

File: composer.json (hunk 2 above)
Comment (AltamashShaikh) [reply composer.json]: Done and it works without publishing the repo

File: .github/workflows/phpcs.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        run:
+          composer --no-plugins config allow-plugins.dealerdirect/phpcodesniffer-composer-installer true -n; 
+          composer config repositories.matomo-coding-standards vcs https://github.com/matomo-org/matomo-coding-standards -n;
+          composer require matomo-org/matomo-coding-standards:dev-master;
+          composer install --dev --prefer-dist --no-progress --no-suggest
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: '7.4'
+          tools: cs2pr
```
Comment (michalkleiner) [inline .github/workflows/phpcs.yml]: I'd swap PHP setup and composer deps install, as you'd normally do things in that order.

File: .gitignore [hunk 5]
```diff
@@ -16,3 +16,11 @@ vendor/**/composer.lock
...
 /vue/dist/*.map
 /vue/dist/*.development.*
+vendor/bin/phpcs
+vendor/bin/phpcbf
+vendor/dealerdirect
+vendor/matomo-org/matomo-coding-standards
+vendor/phpstan
+vendor/slevomat
+vendor/squizlabs
+phpcs-report.xml
```
Comment (michalkleiner) [inline .gitignore]: Do we need to ignore these? We're not committing anything so dirty git status is not an issue, is it?

**Review by snake14 — APPROVED** [review]
Aside from Michal's latest comments, I think things are looking good :+1:

File: .gitignore (hunk 5 above)
Comment (AltamashShaikh) [reply .gitignore]: Its not needed anymore :+1: 
Removed

File: .github/workflows/phpcs.yml (hunk 4 above)
Comment (AltamashShaikh) [reply .github/workflows/phpcs.yml]: @michalkleiner Done

**Comment by AltamashShaikh** [comment]
@sgiehl @michalkleiner Can you guys please check this PR ? If all looks good we can add the similar changes to other plugins.

**Comment by sgiehl** [comment]
I think that looks fine now. :+1:
